File: packages/kits/LocusKit/rust/src/tunnel.rs
//! Tunnel struct. Ports `Tunnel.swift`.
//!
//! A typed cross-reference between two locations in the MemPalace
//! surface. Tunnels link wings, rooms, or specific drawers and are
//! stored directionally so that queries can ask "what does this side
//! know about?" without scanning both endpoints. The symmetric-id
//! contract (canonical id is the hash of the sorted endpoint pair) is
//! documented in `Tunnel.swift` but not enforced at this layer; the
//! enforcement arrives in LOCI-5.
//!
//! Source and target endpoints both carry wing + room + optional
//! drawer id. A `None` drawer id at either end means "the room itself"
//! — useful for room-level concepts that are not anchored to any
//! single drawer.
//!
//! `tombstoned_at` and `removed_by_batch` are present from Rev 1.0 so
//! the schema does not need to migrate when the soft-delete workflow
//! lands.
//!
//! ## Swift-to-Rust shape changes
//!
//! - `Date filedAt` → `i64 filed_at` (epoch seconds). Same convention
//!   as `Drawer::filed_at`; the SQLite column is still TEXT ISO8601.
//! - `Date? tombstonedAt` → `Option<i64> tombstoned_at`.
//! - `TunnelKind = .references` Swift default → Rust callers supply
//!   `TunnelKind::References` explicitly. The bit-layout fallback in
//!   `tunnel_operational.rs` still resolves the default for unknown
//!   raw values.

use crate::tunnel_operational::TunnelKind;

/// A typed cross-reference between two locations.
///
/// Mirrors `Tunnel.swift` field-for-field. The three Int64 bitmaps
/// (`adjective_bitmap`, `operational_bitmap`, `provenance_bitmap`)
/// carry the typed-axis state defined in `adjectives.rs`,
/// `tunnel_operational.rs`, and `provenance.rs` respectively.
/// `Eq` and `Hash` are implemented manually because `order_key: Option<f64>`
/// does not auto-derive those traits. The manual impls use `f64::to_bits()`
/// for bit-exact comparison and hashing, which is correct for storage
/// round-trip equality (NaN-equality is acceptable here because the field
/// never carries NaN in practice).
#[derive(Debug, Clone)]
pub struct Tunnel {
    /// Stable identifier. Conventionally the SHA-256 of the
    /// canonicalised endpoint pair so that A→B and B→A collapse to
    /// one row. This mission accepts whatever id the caller supplies;
    /// LOCI-5 enforces the canonicalisation.
    pub id: String,

    /// Wing of the source endpoint.
    pub source_wing: String,

    /// Room of the source endpoint.
    pub source_room: String,

    /// Drawer id at the source endpoint, when the tunnel targets a
    /// specific drawer. `None` means the room itself.
    pub source_drawer_id: Option<String>,

    /// Wing of the target endpoint.
    pub target_wing: String,

    /// Room of the target endpoint.
    pub target_room: String,

    /// Drawer id at the target endpoint. `None` means the room itself.
    pub target_drawer_id: Option<String>,

    /// Free-form relationship label. Domain-specific; LocusKit does
    /// not validate against a closed catalogue.
    pub label: String,

    /// Typed relationship kind from the closed spec vocabulary
    /// (Appendix A). `kind` is the indexed, finite vocabulary the
    /// retrieval layer dispatches on; `label` is the free-form
    /// human-readable companion.
    pub kind: TunnelKind,

    /// Cross-row adjective bitmap (state, sensitivity, exportability,
    /// trust per spec § 5.5). Stored as a single Int64 column.
    pub adjective_bitmap: i64,

    /// Per-noun operational bitmap (spec § 5.6, tunnel layout).
    /// Accessors in `tunnel_operational.rs` decode direction,
    /// lifecycle, origin_class, strength, and has_inverse.
    pub operational_bitmap: i64,

    /// Provenance bitmap (spec § 5.7, Q1-locked layout). Captures
    /// source type, confirmation, confidence, channel, and
    /// sensitivity at row birth.
    pub provenance_bitmap: i64,

    /// Name of the agent or process that filed this tunnel.
    pub added_by: String,

    /// When the tunnel was added. Epoch seconds in the Rust port;
    /// the SQLite column is TEXT ISO8601 per the fleet rule.
    pub filed_at: i64,

    /// When this tunnel was tombstoned, if it has been. Reserved
    /// for the Rev 2.0 soft-delete workflow.
    pub tombstoned_at: Option<i64>,

    /// Batch identifier used for receipt-based rollback of a
    /// tombstone. Reserved for the Rev 2.0 soft-delete workflow.
    pub removed_by_batch: Option<String>,

    /// Fractional-index ordering key for `Parent` tunnels
    /// (ADR-017 §11). Siblings under the same parent sort by
    /// ascending `order_key`. `None` for non-parent tunnel kinds.
    pub order_key: Option<f64>,
}

impl Tunnel {
    /// Construct a tunnel with all-zero bitmaps and the default
    /// `TunnelKind::References`. Mirrors the Swift designated
    /// initializer's defaulting behavior — callers wanting non-default
    /// bitmaps or a different kind populate the struct fields directly
    /// because Rust does not have argument defaults. The eight
    /// required arguments mirror the Swift initializer's non-default
    /// fields; clippy's too-many-arguments lint is silenced because
    /// shrinking the surface would diverge from the Swift shape.
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        id: String,
        source_wing: String,
        source_room: String,
        target_wing: String,
        target_room: String,
        label: String,
        added_by: String,
        filed_at: i64,
    ) -> Self {
        Tunnel {
            id,
            source_wing,
            source_room,
            source_drawer_id: None,
            target_wing,
            target_room,
            target_drawer_id: None,
            label,
            kind: TunnelKind::References,
            adjective_bitmap: 0,
            operational_bitmap: 0,
            provenance_bitmap: 0,
            added_by,
            filed_at,
            tombstoned_at: None,
            removed_by_batch: None,
            order_key: None,
        }
    }
}

impl PartialEq for Tunnel {
    fn eq(&self, other: &Self) -> bool {
        self.id == other.id
            && self.source_wing == other.source_wing
            && self.source_room == other.source_room
            && self.source_drawer_id == other.source_drawer_id
            && self.target_wing == other.target_wing
            && self.target_room == other.target_room
            && self.target_drawer_id == other.target_drawer_id
            && self.label == other.label
            && self.kind == other.kind
            && self.adjective_bitmap == other.adjective_bitmap
            && self.operational_bitmap == other.operational_bitmap
            && self.provenance_bitmap == other.provenance_bitmap
            && self.added_by == other.added_by
            && self.filed_at == other.filed_at
            && self.tombstoned_at == other.tombstoned_at
            && self.removed_by_batch == other.removed_by_batch
            && self.order_key.map(f64::to_bits) == other.order_key.map(f64::to_bits)
    }
}

impl Eq for Tunnel {}

impl std::hash::Hash for Tunnel {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.id.hash(state);
        self.source_wing.hash(state);
        self.source_room.hash(state);
        self.source_drawer_id.hash(state);
        self.target_wing.hash(state);
        self.target_room.hash(state);
        self.target_drawer_id.hash(state);
        self.label.hash(state);
        self.kind.hash(state);
        self.adjective_bitmap.hash(state);
        self.operational_bitmap.hash(state);
        self.provenance_bitmap.hash(state);
        self.added_by.hash(state);
        self.filed_at.hash(state);
        self.tombstoned_at.hash(state);
        self.removed_by_batch.hash(state);
        self.order_key.map(f64::to_bits).hash(state);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn sample() -> Tunnel {
        Tunnel::new(
            "t-1".to_string(),
            "wing_owner".to_string(),
            "kitchen".to_string(),
            "wing_owner".to_string(),
            "pantry".to_string(),
            "shelves-into".to_string(),
            "alice".to_string(),
            1_700_000_000,
        )
    }

    #[test]
    fn defaults_match_swift_initializer() {
        let t = sample();
        assert_eq!(t.kind, TunnelKind::References);
        assert_eq!(t.adjective_bitmap, 0);
        assert_eq!(t.operational_bitmap, 0);
        assert_eq!(t.provenance_bitmap, 0);
        assert_eq!(t.source_drawer_id, None);
        assert_eq!(t.target_drawer_id, None);
        assert_eq!(t.tombstoned_at, None);
        assert_eq!(t.removed_by_batch, None);
        assert_eq!(t.order_key, None);
    }

    #[test]
    fn fields_round_trip() {
        let mut t = sample();
        t.adjective_bitmap = 42;
        t.operational_bitmap = 17;
        t.provenance_bitmap = 99;
        t.source_drawer_id = Some("d-1".to_string());
        t.target_drawer_id = Some("d-2".to_string());
        t.tombstoned_at = Some(1_700_001_000);
        t.removed_by_batch = Some("batch-1".to_string());

        let cloned = t.clone();
        assert_eq!(cloned, t);
    }
}

File: packages/kits/LocusKit/Sources/LocusKit/Tunnel.swift
import Foundation
import SubstrateKernel

/// A typed cross-reference between two locations in the
/// MemPalace surface.
///
/// Tunnels link wings, rooms, or specific drawers. They are
/// intentionally symmetric in spirit — a tunnel from A to B
/// implies the inverse — but stored directionally so that
/// queries can ask "what does this side know about?" without
/// scanning both endpoints. The symmetric-id contract (the
/// canonical id is a hash of the sorted endpoint pair) is
/// documented here but not enforced at this layer; LOCI-5
/// adds the enforcement once the broader tunnel surface lands.
///
/// Source and target endpoints both carry wing + room + optional
/// drawer id. A nil drawer id at either end means "the room
/// itself" — useful for room-level concepts that are not
/// anchored to any single drawer.
///
/// `tombstonedAt` is written by tunnel-tombstone operations (e.g.
/// outline-parent replacement) and filtered by active-tunnel queries.
/// `removedByBatch` is reserved for the Rev 2.0 receipt-rollback
/// workflow.
public struct Tunnel: Equatable, Hashable, Codable, Sendable {

    /// Stable identifier. Conventionally the SHA-256 of the
    /// canonicalised endpoint pair so that A→B and B→A collapse
    /// to one row. This mission accepts whatever id the caller
    /// supplies; LOCI-5 enforces the canonicalisation.
    public let id: String

    /// Wing of the source endpoint.
    public let sourceWing: String

    /// Room of the source endpoint.
    public let sourceRoom: String

    /// Drawer id at the source endpoint, when the tunnel
    /// targets a specific drawer. Nil means the room itself.
    public let sourceDrawerId: String?

    /// Wing of the target endpoint.
    public let targetWing: String

    /// Room of the target endpoint.
    public let targetRoom: String

    /// Drawer id at the target endpoint. Nil means the room
    /// itself.
    public let targetDrawerId: String?

    /// Free-form relationship label. Domain-specific; LocusKit
    /// does not validate against a closed catalogue.
    public let label: String

    /// Typed relationship kind from the closed spec vocabulary
    /// (Appendix A). Defaulted to `.references` so existing call
    /// sites stay source-compatible. The SQLite `kind_id` column
    /// and the four ALTER guards added by LOCI_V035_05B persist this
    /// value; legacy pre-05B rows fall back to `.references` via the
    /// column's `DEFAULT 1`. Distinct from `label` because `label` is
    /// free-form and `kind` is the indexed, finite vocabulary the
    /// retrieval layer dispatches on.
    public let kind: TunnelKind

    /// Cross-row adjective bitmap (state, sensitivity, exportability,
    /// trust per spec § 5.5). Stored as a single Int64 column. Default
    /// 0 leaves every axis at its zero-value (state=.active,
    /// sensitivity=.normal, exportability=.private, trust=.verbatim).
    public let adjectiveBitmap: Int64

    /// Per-noun operational bitmap (spec § 5.6, tunnel layout).
    /// Stored as a single Int64 column. Accessors in
    /// `TunnelOperational.swift` decode direction, lifecycle,
    /// origin_class, strength, and has_inverse.
    public let operationalBitmap: Int64

    /// Provenance bitmap (spec § 5.7, Q1-locked layout). Stored as a
    /// single Int64 column. Captures source type, confirmation,
    /// confidence, channel, and sensitivity at row birth.
    public let provenanceBitmap: Int64

    /// Name of the agent or process that filed this tunnel.
    public let addedBy: String

    /// When the tunnel was added. TEXT ISO8601 in SQLite.
    public let filedAt: Date

    /// When this tunnel was tombstoned, if it has been. Written by
    /// tunnel-tombstone operations; live tunnel queries filter this field.
    public let tombstonedAt: Date?

    /// Batch identifier used for receipt-based rollback of a
    /// tombstone. Reserved for the Rev 2.0 soft-delete workflow.
    public let removedByBatch: String?

    /// Fractional-index ordering key for `.parent` tunnels
    ///. Siblings under the same parent sort by
    /// ascending `orderKey`. Nil for non-parent tunnel kinds.
    public let orderKey: Double?

    /// Forward-compat JSON extension slot (nullable `ext` column, present
    /// in the tunnels table since the one-`ext`-column-per-persistent-
    /// entity convention landed). MXE-CT3 P2.5 is the first consumer: the
    /// review-ladder endorsement ledger lives here as canonical JSON (see
    /// `TunnelReviewLedger`). Nil for tunnels that have never carried a
    /// review record. Decoded tolerantly from `.json`/`.text` read-back;
    /// unknown keys inside the JSON are preserved on rewrite.
    public let ext: String?

    /// Designated initializer.
    public init(
        id: String,
        sourceWing: String,
        sourceRoom: String,
        sourceDrawerId: String? = nil,
        targetWing: String,
        targetRoom: String,
        targetDrawerId: String? = nil,
        label: String,
        kind: TunnelKind = .references,
        adjectiveBitmap: Int64 = 0,
        operationalBitmap: Int64 = 0,
        provenanceBitmap: Int64 = 0,
        addedBy: String,
        filedAt: Date,
        tombstonedAt: Date? = nil,
        removedByBatch: String? = nil,
        orderKey: Double? = nil,
        ext: String? = nil
    ) {
        self.id = id
        self.sourceWing = sourceWing
        self.sourceRoom = sourceRoom
        self.sourceDrawerId = sourceDrawerId
        self.targetWing = targetWing
        self.targetRoom = targetRoom
        self.targetDrawerId = targetDrawerId
        self.label = label
        self.kind = kind
        self.adjectiveBitmap = adjectiveBitmap
        self.operationalBitmap = operationalBitmap
        self.provenanceBitmap = provenanceBitmap
        self.addedBy = addedBy
        self.filedAt = filedAt
        self.tombstonedAt = tombstonedAt
        self.removedByBatch = removedByBatch
        self.orderKey = orderKey
        self.ext = ext
    }
}

// MARK: - Adjective bitmap accessors

public extension Tunnel {

    /// Decode bits 6–11 of `adjectiveBitmap` as an `AdjectiveSensitivity`.
    ///
    /// Returns `.normal` for unrecognised raw values, matching the estate-level
    /// default access posture (same fail-closed direction as `KGFact.adjectiveSensitivity`
    /// and `Drawer.adjectiveSensitivity`). Named `adjectiveSensitivity` (not `sensitivity`)
    /// to avoid colliding with the provenance-bitmap `sensitivity` accessor. Cookbook
    /// §2.3 6-bit field. Parity peer of Rust `Tunnel::adjective_sensitivity`.
    var adjectiveSensitivity: AdjectiveSensitivity {
        // Cookbook §2.3: sensitivity at bits 6–11 of adjectiveBitmap.
        AdjectiveSensitivity(rawValue: Int(BitField.extractField(adjectiveBitmap, shift: 6, width: 6))) ?? .normal
    }
}
